### ISSUE TITLE
make api stable

### ISSUE THREAD

**Issue by ares89** [Lightning-AI/lightning#17762]
### Description & Motivation

The APIs are changing too frequently and are not backward compatible. When I upgraded from 1.8 to 2.0, I had to rewrite almost all of my code related to Lightning. TensorFlow is like that, but PyTorch is not, which is why PyTorch is more popular.

cc @borda

**Comment by Borda** [comment]
yes, we are high;y committed to API stability from 2.0 onwards.
and apology for the inconvenience you had with updates :rabbit:

**Comment by awaelchli** [comment]
@ares89 Thanks for bringing this up!

Lightning 2.0 marks a new chapter for us. In the [release notes here](https://lightning.ai/pages/releases/2.0.0/), we announced to the community:

> To make the research experience better, we are introducing 2.0:
> No API changes – We commit to backward compatibility in the 2.0 series

We went over the entire API in 1.9.x and overhauled it in 2.0. The core APIs are stable and every public function/method/class will get backward compatibility in the entire 2.x series. In very few cases, we explicitly marked APIs experimental in the docstring, these are the only ones that might change, but only if needed.

If you look at the current changelog for 2.1 for example, which has been in development for several months now, we only currently have 6 API deprecations and 3 removed APIs that are very niche and marked experimental. This list won't really change much until 2.1. These numbers are in very stark contrast to all our previous releases. 

In conclusion, the upgrade process from 2.0 to 2.1 and onward will be an absolute breeze!
I hope this convinces you that we are on a good path and we're happy to hear your thoughts. 

cc @lantiga
